addons/binding/org.openhab.binding.minecraft/src/main/java/org/openhab/binding/minecraft/discovery/MinecraftDiscoveryService.java

```diff
@@ -0,0 +1,263 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.minecraft.discovery;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jmdns.ServiceEvent;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.minecraft.MinecraftBindingConstants;
+import org.openhab.binding.minecraft.config.ServerConfig;
+import org.openhab.binding.minecraft.discovery.scanner.ZeroConfScanner;
+import org.openhab.binding.minecraft.handler.MinecraftServerHandler;
+import org.openhab.binding.minecraft.handler.server.ServerConnection;
+import org.openhab.binding.minecraft.internal.MinecraftHandlerFactory;
+import org.openhab.binding.minecraft.message.data.PlayerData;
+import org.openhab.binding.minecraft.message.data.SignData;
+import org.openhab.binding.minecraft.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import rx.Observable;
+import rx.Subscription;
+import rx.functions.Action1;
+import rx.functions.Func1;
+import rx.functions.Func2;
+import rx.schedulers.Schedulers;
+import rx.subscriptions.CompositeSubscription;
+
+/**
+ * Handles discovery of Minecraft server, players and signs.
+ *
+ * @author Mattias Markehed
+ */
+public class MinecraftDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(MinecraftDiscoveryService.class);
+
+    private static final int DISCOVER_TIMEOUT_SECONDS = 60;
+
+    private CompositeSubscription subscription;
+
+    public MinecraftDiscoveryService() {
+        super(MinecraftBindingConstants.SUPPORTED_THING_TYPES_UIDS, DISCOVER_TIMEOUT_SECONDS, false);
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return MinecraftBindingConstants.SUPPORTED_THING_TYPES_UIDS;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void startScan() {
+        logger.debug("Starting Minecraft discovery scan");
+        discoverServers();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected synchronized void stopScan() {
+        logger.debug("Stopping Minecraft discovery scan");
+        stopServerDiscovery();
+        super.stopScan();
+    }
+
+    /**
+     * Setup subscribers searching for servers.
+     * Start scanning for items on servers.
+     */
+    private void discoverServers() {
```

Comment:
This discovery service should not implement the discovery of servers, but only of Players and Signs.
For the servers, you should indeed create a new service that implements an MDNSDiscoveryParticipant, see https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html#mdns-discovery. This will also remove the need to use jmdns directly.

Note that the discovery service for players and signs depends on the server (so you might have multiple instances of such discovery services). Therefore it must not be declared through DS (an xml file), but should be registered in the handler factory, whenever a handler for a server is created, see e.g. [this example](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java#L86).